### PR TITLE
[FsxS3Access] Add missing permissions needed to import/export from S3 when using Fsx for Lustre via API

### PR DESCRIPTION
**Pull request by EddyMM**
### Description of changes
- Creation of a cluster with FsxLuster storage via ParallelCluster API fails due to the user role missing some permissions needed to import/export from S3.
- This commit creates an inline-policy with the needed permissions and associates it with the user role.
- The inline-policy is only created if the API stack is created with certain parameters (EnableFSxS3Access and FsxS3Buckets).

### Tests
* Manually tested by first reproducing the error through API Gateway
* Applied the changes in this commit to a copy of the current CFN Template
* Recreated the API and verified that the `FSxS3AccessPolicy` is created and associated with the `ParallelClusterUserRole`
* Made an API request using API Gateway to create a cluster
* Got a 202 OK response status with these results

```
{
  "cluster": {
    "cloudformationStackArn": "<STACK_ARN>",
    "cloudformationStackStatus": "CREATE_IN_PROGRESS",
    "clusterName": "<CLUSTER_NAME>",
    "clusterStatus": "CREATE_IN_PROGRESS",
    "region": "eu-west-1",
    "version": "3.1.4"
  }
}
```

### References
* [CloudFormation Conditions](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-condition.html)

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [ ] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [x] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -40,6 +40,7 @@ CHANGELOG
 - Fix default for disable validate and test components when building custom AMI. The default was to disable those components, but it wasn't effective.
 - Handle corner case in the scaling logic when instance is just launched and the describe instances API doesn't report yet all the EC2 info.
 - Dropped validation that would prevent ARM instance type to be used when `DisableSimultaneousMultithreading` was set to true.
+- Fix S3BucketUriValidator error that occurs when creating a cluster with FsxStorage using ParallelCluster API
```
Comment (gmarciani) [inline CHANGELOG.md]: What about "Add missing permissions needed to import/export from S3 when using Fsx for Lustre via ParallelCluster API."?

File: api/infrastructure/parallelcluster-api.yaml [hunk 2]
```diff
@@ -59,6 +59,23 @@ Parameters:
       - true
       - false
 
+  EnableFSxS3Access:
+    Description: |
+      When set to true the ParallelCluster API can access, write to the S3 buckets specified in the Filed FsxS3Bucket, it is needed to import/export from/to S3 when creating an FSx filesystem.
+      WARNING - setting this to true grants to the Lambda function S3 Get*, List* and PutObject privileges on the buckets specified in FsxS3Buckets.
```
Comment (gmarciani) [inline api/infrastructure/parallelcluster-api.yaml]: Let's replace "WARNING - setting ..." with "NOTE - Setting ..."

File: api/infrastructure/parallelcluster-api.yaml [hunk 3]
```diff
@@ -59,6 +59,23 @@ Parameters:
...
+    Type: String
+    Default: false
+    AllowedValues:
+      - true
+      - false
+
+  FsxS3Buckets:
+    Description: |
+      Comma separated arn of the S3 buckets, to allow the lambda function to import/export from/to S3 when creating an FSx filesystem.
```
Comment (gmarciani) [inline api/infrastructure/parallelcluster-api.yaml]: Comma separated list of S3 bucket ARNs

File: api/infrastructure/parallelcluster-api.yaml [hunk 4]
```diff
@@ -59,6 +59,23 @@ Parameters:
...
+      WARNING - The setting is used only when EnableFSxS3Access is set to true. (example arn:aws:s3:::<S3_BUCKET_1>,arn:aws:s3:::<S3_BUCKET_2>)
```
Comment (gmarciani) [inline api/infrastructure/parallelcluster-api.yaml]: 1. Let's replace WARNING with NOTE.
!. The example should rather be before the NOTE

File: api/infrastructure/parallelcluster-api.yaml [hunk 5]
```diff
@@ -59,6 +59,23 @@ Parameters:
...
+    Type: String
+    AllowedPattern: ^($|arn:.*:s3:[a-z0-9\-]*:([0-9]{12})*:.+)
```
Comment (gmarciani) [inline api/infrastructure/parallelcluster-api.yaml]: Where did you get this regex?
Be careful that it matches strings such as "arn:aws:s3:::S3_BUCKET_1,   arn:aws:s3:::S3_BUCKET_2", (with blank spaces), that are not split as expected within the policy statement

File: tests/iam_policies/user-role.cfn.yaml [hunk 6]
```diff
@@ -20,12 +20,31 @@ Parameters:
     AllowedValues:
       - true
       - false
+  EnableFSxS3Access:
```
Comment (gmarciani) [inline tests/iam_policies/user-role.cfn.yaml:23]: Let's align this template to the changes in parallelcluster-api.yaml

File: api/infrastructure/parallelcluster-api.yaml (hunk 5 above)
Comment (EddyMM) [reply api/infrastructure/parallelcluster-api.yaml]: It's a custom regex. I'll test it against such variations of S3 arns.

File: CHANGELOG.md [hunk 7]
```diff
@@ -40,6 +40,7 @@ CHANGELOG
 - Fix default for disable validate and test components when building custom AMI. The default was to disable those components, but it wasn't effective.
 - Handle corner case in the scaling logic when instance is just launched and the describe instances API doesn't report yet all the EC2 info.
 - Dropped validation that would prevent ARM instance type to be used when `DisableSimultaneousMultithreading` was set to true.
+- Add missing permissions needed to import/export from S3 when using Fsx for Lustre via ParallelCluster API.
```
Comment (gmarciani) [inline CHANGELOG.md]: Minor: Fsx should be FSx.

File: api/infrastructure/parallelcluster-api.yaml [hunk 8]
```diff
@@ -459,6 +483,31 @@ Resources:
...
+              - iam:AttachRolePolicy
+              - iam:PutRolePolicy
+            Resource: !Sub arn:${AWS::Partition}:iam::${AWS::AccountId}:role/aws-service-role/s3.data-source.lustre.fsx.amazonaws.com/*
+            Effect: Allow
+            Sid: FSxS3PoliciesAttach
+          - Action:
+              - s3:Get*
+              - s3:List*
+              - s3:PutObject
+            Resource: !Split [",", !Ref FsxS3Buckets]
```
Comment (gmarciani) [inline api/infrastructure/parallelcluster-api.yaml]: The `s3:PutObject` requires the resource to be BUCKET_NAME/*, doesn't it?

Comment (EddyMM) [reply api/infrastructure/parallelcluster-api.yaml]: Yes, I'll run some additional tests to confirm. This can be specified in the FsxS3Buckets parameter but perhaps we may want to implicitly add the `*` wildcard.

Comment (EddyMM) [reply api/infrastructure/parallelcluster-api.yaml]: Good catch, i've updated the parameter pattern to make sure the user provides a list with bucket ARNs that have `/*`

File: api/infrastructure/parallelcluster-api.yaml [hunk 9]
```diff
@@ -459,6 +483,41 @@ Resources:
...
+              - ","
+              - !Join # "abc,xyz,qrs,abc/*,xyz/*,qrs/*"
+                - ""
+                - - !Ref FsxS3Buckets
+                  - ","
+                  - !Join # "abc/*,xyz/*,qrs/*"
+                    - "/*,"
+                    - !Split # ["abc", "xyz" , "qrs/*]"
+                      - ","
+                      - !Join ["", [!Ref FsxS3Buckets, "/*"]] ### "abc,xyz,qrs/*"
```
Comment (demartinofra) [inline api/infrastructure/parallelcluster-api.yaml]: This looks good but I'm wondering if we can leverage !Sub to make it more readable. For example something like:
```
!Sub
- "${FsxS3Buckets},${FsxS3BucketsObjects}"
- FsxS3BucketsObjects: !Join ["/*,", !Split [",", !Sub "${FsxS3Buckets}/*"]]
```

syntax to be checked :)

**Review by demartinofra — COMMENTED** [review]
LGTM, only left a minor comment on a potential improvement

File: api/infrastructure/parallelcluster-api.yaml (hunk 9 above)
Comment (EddyMM) [reply api/infrastructure/parallelcluster-api.yaml]: I'll take a look into it. `!Sub` would definitely help simplify it.

File: api/infrastructure/parallelcluster-api.yaml [hunk 10]
```diff
@@ -459,6 +484,37 @@ Resources:
...
+              - s3:Get*
+              - s3:List*
+              - s3:PutObject
+            Resource: !Split
+              - ","
+              - !If [
+                UseAllBucketsForFSxS3,
+                "*",
+                !Sub ["${FsxS3Buckets},${FsxS3BucketsObjects}", FsxS3BucketsObjects: !Join ["/*,", !Split [",", !Sub "${FsxS3Buckets}/*"]]]
+                ]
```
Comment (demartinofra) [inline api/infrastructure/parallelcluster-api.yaml]: nit: for the !If since you are not inlining you could use `-` instead of `[` which is more yaml friendly.

**Review by demartinofra — APPROVED** [review]
